### PR TITLE
ARROW-4752: [Rust] Add explicit SIMD vectorization for the divide kernel

### PR DESCRIPTION
**Pull request by paddyhoran**
Adds SIMD support for divide binary op.  There are probably a few more optimizations to be made but the performance of divide is now very close to the other binary ops on my computer which is pretty good considering all the extra checking involved to allow for divide by zero correctly.

PTAL

### CODE REVIEW

File: rust/arrow/src/compute/kernels/arithmetic.rs [hunk 1]
```diff
@@ -121,6 +122,71 @@ where
...
+                (result.data_mut().as_mut_ptr() as *mut T::Native).offset(i as isize),
+                lanes,
+            )
+        };
+        T::write(simd_result, result_slice);
+    }
+
+    let null_bit_buffer = bitmap.and_then(|b| Some(b.bits));
+
+    let data = ArrayData::new(
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/arithmetic.rs]: @sunchao has previously stated preference for us using a builder, but I'm fine with this

**Review by nevi-me — APPROVED** [review]
🎉🎉🎉

File: rust/arrow/src/compute/kernels/arithmetic.rs [hunk 2]
```diff
@@ -197,8 +263,13 @@ where
         + Sub<Output = T::Native>
         + Mul<Output = T::Native>
         + Div<Output = T::Native>
-        + Zero,
+        + Zero
+        + One,
 {
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    return simd_divide(&left, &right);
```
Comment (liurenjie1024) [inline rust/arrow/src/compute/kernels/arithmetic.rs]: Is it more readable to use the 
`if cfg!`
pattern?

File: rust/arrow/src/compute/kernels/arithmetic.rs [hunk 3]
```diff
@@ -274,6 +345,19 @@ mod tests {
         assert_eq!(9, c.value(4));
     }
 
+    #[test]
+    fn test_primitive_array_divide_with_nulls() {
```
Comment (liurenjie1024) [inline rust/arrow/src/compute/kernels/arithmetic.rs]: I would suggest to use template to covert all numeric types in test.

**Review by liurenjie1024 — APPROVED** [review]
LGTM. Just some minor changes.

File: rust/arrow/src/compute/kernels/arithmetic.rs (hunk 3 above)
Comment (paddyhoran) [reply rust/arrow/src/compute/kernels/arithmetic.rs]: @liurenjie1024 I'm not sure what you mean, could you explain?

Comment (liurenjie1024) [reply rust/arrow/src/compute/kernels/arithmetic.rs]: I mean maybe we should cover all numerics types, not just int32?

**Review by andygrove — APPROVED** [review]
LGTM!

File: rust/arrow/src/compute/kernels/arithmetic.rs (hunk 1 above)
Comment (paddyhoran) [reply rust/arrow/src/compute/kernels/arithmetic.rs]: So I remember why I didn't use the builder now.  `null_bit_buffer` is an `Option<Bitmap>`.  There is no method on the builder that accepts an `Option<Bitmap>` only one that accepts `Bitmap` which leads to a match on `null_bit_buffer` which is not very clean.  `new` accepts the `Option<Bitmap>` directly.

File: rust/arrow/src/compute/kernels/arithmetic.rs (hunk 2 above)
Comment (paddyhoran) [reply rust/arrow/src/compute/kernels/arithmetic.rs]: Follow up JIRA is [here](https://issues.apache.org/jira/browse/ARROW-6408?filter=-2).

File: rust/arrow/src/compute/kernels/arithmetic.rs (hunk 3 above)
Comment (paddyhoran) [reply rust/arrow/src/compute/kernels/arithmetic.rs]: Ok, yep good point.  Follow up JIRA is [here](https://issues.apache.org/jira/browse/ARROW-6409?filter=-2)

**Pull request closed without merging**